service/project.go

```diff
@@ -211,6 +219,147 @@ func (s *ProjectService) DeleteEnvironment(ctx context.Context, projectID, envID
+	if err := s.authSvc.AuthorizeAdminRole(ctx, authUserID); err != nil {
+		return model.ProjectInvitation{}, err
+	}
+
+	if _, err := s.Get(ctx, c.ProjectID, authUserID); err != nil {
+		return model.ProjectInvitation{}, err
+	}
+
+	tkn, err := cryptox.NewToken()
+	if err != nil {
+		return model.ProjectInvitation{}, err
+	}
+
+	i, err := model.NewProjectInvitation(c, tkn)
+	if err != nil {
+		return model.ProjectInvitation{}, apierrors.NewProjectInvitationUnprocessableError().Wrap(err).WithMessage(err.Error())
+	}
+
+	// TODO check if the user is already a member of the project
+
+	if exists, err := s.invitationExists(ctx, i.Email, c.ProjectID); err != nil {
+		return model.ProjectInvitation{}, err
+	} else if exists {
+		return model.ProjectInvitation{}, apierrors.NewProjectInvitationAlreadyExistsError()
+	}
+
+	if err := s.invitationRepo.Create(ctx, i); err != nil {
+		return model.ProjectInvitation{}, err
+	}
+
+	// TODO send invitation email
+
+	return i, nil
+}
+
+func (s *ProjectService) GetInvitations(ctx context.Context, projectID, authUserID uuid.UUID) ([]model.ProjectInvitation, error) {
+	if err := s.authSvc.AuthorizeAdminRole(ctx, authUserID); err != nil {
+		return nil, err
+	}
+
+	_, err := s.Get(ctx, projectID, authUserID)
+	if err != nil {
+		return nil, err
+	}
+
+	i, err := s.invitationRepo.ReadAllForProject(ctx, projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := model.ValidateProjectInvitations(i); err != nil {
+		return nil, apierrors.NewProjectInvitationCorruptedError().Wrap(err)
+	}
+
+	return i, nil
+}
+
+func (s *ProjectService) DeleteInvitation(ctx context.Context, projectID, invitationID, authUserID uuid.UUID) error {
+	if err := s.authSvc.AuthorizeAdminRole(ctx, authUserID); err != nil {
+		return err
+	}
+
+	_, err := s.Get(ctx, projectID, authUserID)
+	if err != nil {
+		return err
+	}
+
+	i, err := s.invitationRepo.Read(ctx, invitationID)
+	if err != nil {
+		switch {
+		case dberrors.IsNotFoundError(err):
+			return apierrors.NewProjectInvitationNotFoundError().Wrap(err)
+		default:
+			return err
+		}
+	}
+
+	return s.invitationRepo.Delete(ctx, i.ID)
+}
+
+func (s *ProjectService) AcceptInvitation(ctx context.Context, tkn cryptox.Token) error {
+	invitation, err := s.getPendingInvitationByToken(ctx, tkn)
+	if err != nil {
+		return err
+	}
+
+	/*
+		When an invitation is confirmed, db function handle_confirmed_invitation is triggered.
+		The function checks if the email is already associated with a user account.
+		If user exists, the function creates a project member and deletes the invitation.
+
+		When user signs up, function handle_new_user is triggered.
+		The function checks if the user's email is associated with a confirmed invitation(s).
+		If so, the function creates a project member(s) and deletes the invitation(s).
+
+		This approach is used because current repository implementation does not support transactions.
+	*/
+	invitation.Accept()
```

Comment:
The reason for `accepted_awaiting_registration` status.

Since Supabase doesn't support transactions and I don't control the user signing process via Google SSO, setting up database triggers is the best solution.

It will be implemented together with project members.

Reply:
Aha, so that's because Supabase doesn't support transactions..

And could you explain more the Google SSO argument?

Other than that, probably OK, looks interesting.

Although, I'd say.. if you have time, I'd think whether not having transactions will introduce a really bad, inconsistent state.
If it's only about an extremely minimal chance of having a very unharmful state, then in the long run, consider moving everything to Go and re-ordering everything so that the number of really dangerous inconsistent states is minimized...

Reply:
Another comment was, I will try to answer all of them it this thread 

> In general, I'm curious about db triggers, but..
>Apart from being detached from this, core place of business logic..
> Is it necessary?
> How will the debugging work?
> Do you check for example if invitation already has status accepted_... or other potentially invalid states?..

Reply:
> And could you explain more the Google SSO argument?

When a user signs up via Google SSO using Supabase, my backend API plays no role (it is not called at all). Therefore, the only way I can perform some action upon user registration is to set up a trigger in the database.

> Although, I'd say.. if you have time, I'd think whether not having transactions will introduce a really bad, inconsistent state.

If I move all logic to Go, I would:

1. Check if the user already has an account (1st query).

-> User has account
2. Delete the invitation (2nd query).
3. Create a project member (3rd query).

-> User has no account yet
2. Update invitation status (2nd query)

If something fails along the way (or changes), I would end up in an inconsistent state. 

>  Do you check for example if invitation already has status accepted_... or other potentially invalid states?..

Yes, I would check all these cases in a PostgreSQL function.

---

I agree with you that a downside of the solution is its testability and the fact that such core business logic is handled only in the database layer.

One of the objectives of my thesis is to evaluate the efficiency of using Supabase (and its REST API for database access instead of direct database connections). And I will mention all those downsides.

I have already spent quite lot of time debugging unofficial Supabase client for Go (there is no official yet). If this evaluation was not a part of my thesis goals, I would consider connecting to db directly.

Reply:
Alternatively, instead of setting up a trigger, I can directly call a PostgreSQL function from the repository.

Reply:
Aha, OK, makes sense.

But still:
>If something fails along the way (or changes), I would end up in an inconsistent state.

What _exactly_ can go wrong?

For example if the account exists:
1. The account existence query is  (especially since it's such a rare, non-critical case in a non-real-time application).
2. You can rather first create an account, and then delete an invitation, in one service call. What can go wrong - during or after the project member creation the user will accept the invitation again? Weird, non-critical corner case, but you have a uniqueness constraint on `(project_id, email)`, so it will fail, you'll just have to handle this uniqueness violation error and wrap it into smth like `REPEATED_INVITATION_ACC_ERROR`...

Or it might be interesting to explore whether we can have even more consistency/integrity imposed declaratively, i.e. some references between the tables, maybe a new table, uniqueness constraints...

But fair enough, for research purposes, indeed, it might be interesting to see how helpful or dragging the db functions/triggers approach will be.

>Alternatively, instead of setting up a trigger, I can directly call a PostgreSQL function from the repository.

Yes, that sounds like a good compromise.
Although, you already have a comment there, so I guess we can keep it as it is.

Reply:
Ok, thanks for comments! Anyway I am going to implement it in the next PR so I will consider all the options again. 👍